src/core/qgsprojectservervalidator.cpp

```diff
@@ -0,0 +1,132 @@
+ ***************************************************************************/
+
+/***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+
+#include "qgsapplication.h"
+#include "qgslayertreelayer.h"
+#include "qgsprojectservervalidator.h"
+#include "qgsvectorlayer.h"
+
+
+QString QgsProjectServerValidator::displayValidationError( QgsProjectServerValidator::ValidationError error )
+{
+  switch ( error )
+  {
+    case QgsProjectServerValidator::Encoding:
+      return QObject::tr( "Encoding is not set properly. A non 'System' encoding is required" );
+    case QgsProjectServerValidator::ShortNames:
+      return QObject::tr( "Layer short name is not valid. It must start with an unaccented alphabetical letter, followed by any alphanumeric letters, dot, dash or underscore" );
+    case QgsProjectServerValidator::DuplicatedNames:
+      return QObject::tr( "One or more layers or groups have the same name or short name. Both the 'name' and 'short name' for layers and groups needs to be unique" );
+  }
+  return QString();
+}
+
+void QgsProjectServerValidator::browseLayerTree( QgsLayerTreeGroup *treeGroup, QStringList &owsNames, QStringList &encodingMessages )
+{
+  QList< QgsLayerTreeNode * > treeGroupChildren = treeGroup->children();
+  for ( int i = 0; i < treeGroupChildren.size(); ++i )
+  {
+    QgsLayerTreeNode *treeNode = treeGroupChildren.at( i );
+    if ( treeNode->nodeType() == QgsLayerTreeNode::NodeGroup )
+    {
+      QgsLayerTreeGroup *treeGroupChild = static_cast<QgsLayerTreeGroup *>( treeNode );
+      QString shortName = treeGroupChild->customProperty( QStringLiteral( "wmsShortName" ) ).toString();
+      if ( shortName.isEmpty() )
+        owsNames << treeGroupChild->name();
+      else
+        owsNames << shortName;
+      browseLayerTree( treeGroupChild, owsNames, encodingMessages );
+    }
+    else
+    {
+      QgsLayerTreeLayer *treeLayer = static_cast<QgsLayerTreeLayer *>( treeNode );
+      QgsMapLayer *layer = treeLayer->layer();
+      if ( layer )
+      {
+        QString shortName = layer->shortName();
+        if ( shortName.isEmpty() )
+          owsNames << layer->name();
+        else
+          owsNames << shortName;
+
+        if ( layer->type() == QgsMapLayerType::VectorLayer )
+        {
+          QgsVectorLayer *vl = static_cast<QgsVectorLayer *>( layer );
+          if ( vl->dataProvider()->encoding() == QLatin1String( "System" ) )
```

Comment:
Not sure if that's somehow prevented by the context already

```suggestion
          if ( vl->dataProvider() && vl->dataProvider()->encoding() == QLatin1String( "System" ) )
```